### PR TITLE
Omnipool cex arbitrage version 2

### PR DESCRIPTION
**Pull request by poliwop**

### CODE REVIEW

File: hydradx/model/amm/omnipool_amm.py [hunk 1]
```diff
@@ -1021,47 +1021,47 @@ def __init__(self, state: OmnipoolState):
...
+def asset_invariant(state: OmnipoolState, i: str) -> float:
     """Invariant for specific asset"""
-    return self.liquidity[i] * self.lrna[i]
+    return state.liquidity[i] * state.lrna[i]
 
 
-def swap_lrna_delta_Qi(self: OmnipoolState, delta_ri: float, i: str) -> float:
-    return self.lrna[i] * (- delta_ri / (self.liquidity[i] + delta_ri))
+def swap_lrna_delta_Qi(state: OmnipoolState, delta_ri: float, i: str) -> float:
+    return state.lrna[i] * (- delta_ri / (state.liquidity[i] + delta_ri))
```
Comment (jepidoptera) [inline hydradx/model/amm/omnipool_amm.py:1056]: I'm realizing that the only reference to this function anywhere in the project is in a test, which just tests this function. We could get rid of the function and the test and nothing else would notice.

File: hydradx/model/amm/omnipool_amm.py [hunk 2]
```diff
@@ -1021,47 +1021,47 @@ def __init__(self, state: OmnipoolState):
...
 
-def swap_lrna_delta_Ri(self: OmnipoolState, delta_qi: float, i: str) -> float:
-    return self.liquidity[i] * (- delta_qi / (self.lrna[i] + delta_qi))
+def swap_lrna_delta_Ri(state: OmnipoolState, delta_qi: float, i: str) -> float:
+    return state.liquidity[i] * (- delta_qi / (state.lrna[i] + delta_qi))
 
 
-def weight_i(self: OmnipoolState, i: str) -> float:
-    return self.lrna[i] / self.lrna_total
+def weight_i(state: OmnipoolState, i: str) -> float:
```
Comment (jepidoptera) [inline hydradx/model/amm/omnipool_amm.py:1063]: Same as above goes for this function.

**Review by jepidoptera — APPROVED** [review]

File: hydradx/model/amm/arbitrage_agent.py [hunk 3]
```diff
@@ -162,13 +150,21 @@ def calculate_arb_amount_bid(
...
+                # best trade will involve trading max_liquidity, so can be calculated as a sell
+                test_agent = agent.copy()
+                test_state = state.copy()
+                test_state.swap(test_agent, tkn_buy=tkn, tkn_sell=numeraire, sell_quantity=max_liq_num)
+                return test_agent.holdings[tkn] - agent.holdings[tkn]
+            else:
+                amt_low = amt
+                best_buy_spot = buy_spot
+
+        if amt_high == amt_low:  # full amount can be traded
```
Comment (jepidoptera) [inline hydradx/model/amm/arbitrage_agent.py:166]: 👍

File: hydradx/model/processing.py [hunk 4]
```diff
@@ -191,10 +191,56 @@ def import_monthly_binance_prices(
     return price_data
 
 
-def parse_kraken_orderbook(orderbook):
-    asks = [{'price': float(ask[0]), 'amount': float(ask[1])} for ask in orderbook['asks']]
-    bids = [{'price': float(bid[0]), 'amount': float(bid[1])} for bid in orderbook['bids']]
-    return {'bids': bids, 'asks': asks}
+def get_kraken_orderbook(tkn_pair: tuple, orderbook_url: str) -> OrderBook:
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/processing.py:194]: I will do this in my new branch

File: hydradx/model/amm/omnipool_amm.py (hunk 1 above)
Comment (poliwop) [reply hydradx/model/amm/omnipool_amm.py:1056]: Make a ticket to do this separately, I don't wanna do this as part of this arb update.

That said I'm a little surprised, do we get the delta_Qi of a LRNA swap some other way? Or is it only necessary when buying LRNA or something which we don't currently have implemented?

File: hydradx/model/amm/omnipool_amm.py (hunk 2 above)
Comment (poliwop) [reply hydradx/model/amm/omnipool_amm.py:1063]: I think this function is worth having

**Pull request merged**